cmd/erasure-object.go

```diff
@@ -1470,7 +1470,17 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 			actualSize = n
 		}
 	}
-	if fi.Checksum == nil {
+	// If ServerSideChecksum is wanted for this object, it takes precedence
+	// over opts.WantChecksum.
+	if opts.WantServerSideChecksumType.IsSet() {
+		serverSideChecksum := r.RawServerSideChecksumResult()
+		if serverSideChecksum != nil {
+			fi.Checksum = serverSideChecksum.AppendTo(nil, nil)
+			if opts.EncryptFn != nil {
+				fi.Checksum = opts.EncryptFn("object-checksum", fi.Checksum)
+			}
+		}
+	} else if fi.Checksum == nil && opts.WantChecksum != nil {
```

Comment:
@klauspost This is the only line that concerned me a little bit, as it's different in AIStor. 

We weren't checking `if fi.Checksum == nil` in AIStor before, but it is being checked here. This looks like it might be set if `ReplicationSsecChecksumHeader` is set which was introduced by your PR here: https://github.com/minio/minio/pull/19896/files

Thoughts on this?